Embed yarn editor in GD

blurymind opened the pull request:
This bundles Yarn - dialogue tree editor with gdevelop - making it possible to create and edit dialogue trees directly inside gdevelop.
![yarnGD](https://user-images.githubusercontent.com/6495061/62571206-b730c300-b888-11e9-810e-a21a5cb8dd33.png)

It's worth noting that Yarn's devs gave me full access to do commits to it. If there is anything required to fix it- I can fix it :) 

This design bundles a new version of the electron branch of yarn, which I did some refactoring to in order to make it also work as a standalone web app. I am attaching it as a zip here, for those who want to test. Here is a build of
https://github.com/InfiniteAmmoInc/Yarn/tree/electron 

updated yarn package:
https://github.com/YarnSpinnerTool/YarnEditor/commit/a5fb76b771481c92b2ec4af47d4da460c4b56aa8
[yarn-editor.zip](https://github.com/4ian/GDevelop/files/3627667/yarn-editor.zip) (a5fb76b )



This pr also gives the path editor a new feature- a [>] button to toggle it, as well as css cleanup - to make the style easier to read and carry over to other editors.

Not sure if I should do the bbs parser next or ability to read mod music files :P 

updated the example:
[dialogueExample.zip](https://github.com/4ian/GDevelop/files/3513105/dialogueExample.zip)
Here is a screenshot example of how to set up an entire npc system with the extension:
![2019-08-18_15-00](https://user-images.githubusercontent.com/6495061/63225553-09e37680-c1c9-11e9-8a3d-34a08a8bda51.png)
![2019-08-18_15-17](https://user-images.githubusercontent.com/6495061/63225727-65aeff00-c1cb-11e9-994c-85adaae1ed7d.png)

One with animated avatars and conditional branches

trello
https://trello.com/c/vcRqzoGk/74-bundle-yarn-a-dialogue-editor-for-interactive-stories

4ian commented on newIDE/app/src/ResourcesList/LocalYarnBridge.js:
```diff
@@ -0,0 +1,53 @@
+
+  const externalEditorData = {
+    resourcePath: initialResourcePath,
+    // $FlowFixMe - TODO: There is an error here to be solved.
```
Can you give me the error that is sent here?

blurymind replied:
this is a bit of copy pasta from the one jfxr uses. Not sure if we actually need the flowfixme here tbh, but will test

blurymind replied:
there is actually a flow error if I remove the flowfixme:
```
Error ┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈ src/ResourcesList/LocalYarnBridge.js:46:23

Cannot call onChangesSaved with array literal bound to the first parameter because:
 • property name is missing in object literal [1] but exists in object type [2] in array element.
 • property originalIndex is missing in object literal [1] but exists in object type [2] in array element.

     src/ResourcesList/LocalYarnBridge.js
     43│       const resourceName = path.relative(projectPath, newFilePath);
     44│       createOrUpdateResource(project, new gd.JsonResource(), resourceName);
     45│
 [1] 46│       onChangesSaved([{}], resourceName);
     47│     }
     48│   );
     49│

     src/ResourcesList/ResourceExternalEditor.flow.js
 [2] 14│     Array<{
     15│       path?: string,
     16│       name: string,
     17│       originalIndex: ?number,
     18│       metadata?: ?Object,
     19│     }>,

```

4ian replied:
See my other comment, if you pass an empty array you should be able to remove this FlowFixMe.
If there is still an error, take a look at it, but there is no reason we need to add FlowFixMe - it's basically a hack and remove the value of having Flow checking the types.

4ian commented on newIDE/app/public/external/yarn/yarn-main.js:
```diff
@@ -0,0 +1,77 @@
+// Called to load yarn data. Should be called after the window is fully loaded.
+
+ipcRenderer.on('yarn-open', (event, receivedOptions) => {
+  if (!yarn) return;
```
Maybe we should log an error if this return is called? Looks like something that should never happen right, in theory?

blurymind replied:
i will probably remove this. You are right, since I am using a callback - this will never happen

4ian commented on newIDE/app/public/external/yarn/yarn-main.js:
```diff
@@ -0,0 +1,77 @@
+
+  //Make a header
+  const pathEditorHeaderDiv = document.getElementById('path-editor-header');
+  const pathControl = createPathEditorHeader({
```
Nitpicking: the function is called "createPathEditorHeader" so can we call the variable "pathEditorHeader"?

4ian commented on newIDE/app/public/external/yarn/README.md:
```diff
@@ -0,0 +1,7 @@
+This folder contains sources to embed Yarn editor (https://github.com/InfiniteAmmoInc/Yarn) so that it can
+be used directly from GDevelop to edit/create Dialogue Trees.
+
+Yarn sources are downloaded by `import-zipped-editor.js` script. They are raw, unchanged sources
```
Thanks for the README :+1:

blurymind replied:
no probs, just following the convention established with the other ones :)

4ian commented on newIDE/app/public/external/yarn/.gitignore:
```diff
@@ -0,0 +1,87 @@
+.idea/
```
This .gitignore is good but a bit too complete, let's instead put only the things useful for yarn (other things like .idea/grun/OSX/thumbnails should in theory be handled, if necessary, by the root .gitignore of the GD repo)

blurymind replied:
no probs, just following the convention established with the other ones :)

blurymind replied:
this file is actually not needed. It was autocreated by vscode I think. The gitignore is handled elsewhere, so I will delete this

4ian commented:
Looks pretty nice! :smile:  

> I will get these new changes merged to yarn's electron branch soon too.

Ok cool, would be great if the yarn-editor.zip could be generated from a well known branch/commit so that it's easily reproducible in case me or someone else has to make a fix in Yarn and re-bundle it.

Thanks for taking the time to refactor things related to the header, I'll give a more in depth look at the code.

Would it be possible to have a very very simple example file: a game, with a simple dialogue json file, a text object and just a few events to kickstart the dialog and maybe a key to press (or click) to continue.
Context: 
* This will allow people to quickly see how to use this extension and try it.
* For me that don't know well how Yarn or the extension work, it's a way to quickly test that the extension still work after I do a change/refactoring to the extension or Yarn.

In a way, consider the example(s) as your guarantee that I won't break your extension behind your back when you're not looking at it ;)

blurymind commented:
Yes I plan to do it this week some time. Just want to make sure everything
is Rock solid with it in GD, as an electron app and as a webapp. The webapp
support is going to be nice, since people will be able to use it straight
in the browser from a smartphone. It will also be a step towards having
yarn behave really well with mobile screens

On Wed, Aug 7, 2019, 10:17 AM Florian Rival <notifications@github.com>
wrote:

> Looks pretty nice! 😄
>
> I will get these new changes merged to yarn's electron branch soon too.
>
> Ok cool, would be great if the yarn-editor.zip could be generated from a
> well known branch/commit so that it's easily reproducible in case me or
> someone else has to make a fix in Yarn and re-bundle it.
>
> Thanks for taking the time to refactor things related to the header, I'll
> give a more in depth look at the code.
>
> Would it be possible to have a very very simple example file: a game, with
> a simple dialogue json file, a text object and just a few events to
> kickstart the dialog and maybe a key to press (or click) to continue.
> Context:
>
>    - This will allow people to quickly see how to use this extension and
>    try it.
>    - For me that don't know well how Yarn or the extension work, it's a
>    way to quickly test that the extension still work after I do a
>    change/refactoring to the extension or Yarn.
>
> In a way, consider the example(s) as your guarantee that I won't break
> your extension behind your back when you're not looking at it ;)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/4ian/GDevelop/pull/1185?email_source=notifications&email_token=ABRRWVJF773V4IGCGREQQN3QDKHI3A5CNFSM4IJZRQ32YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD3XYTUQ#issuecomment-519014866>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABRRWVIKOZSAYFSPRBAVONDQDKHI3ANCNFSM4IJZRQ3Q>
> .
>

blurymind commented:
I applied the review notes and did some cleanup. Also committed the latest changes to yarn's git, so you can see that the currently attached build is based on Latest commit
7aab340 of the electron branch here
https://github.com/InfiniteAmmoInc/Yarn/tree/electron

4ian commented:
Cool :) I have to take a look at testing this. How stable do you think this is?

blurymind commented:
> Cool :) I have to take a look at testing this. How stable do you think this is?

I am confident it is pretty stable in terms of editing/saving :) I am further testing the runtime (extension) in a more elaborate demo project atm. 

I was wondering if it would be useful to add the name of the editor as metadata of the resource. It currently is not adding metadata

4ian commented:
> I was wondering if it would be useful to add the name of the editor as metadata of the resource. It currently is not adding metadata

Yes please :) Will be useful when we start having more than one editor using json resources.

blurymind commented:
@4ian I found a bug in my dialogue extension. Will you be ok if I include it in this pr?

4ian commented:
Yup it's fine as related :)

On Tue, 13 Aug 2019, 13:28 Todor Imreorov, <notifications@github.com> wrote:

> @4ian <https://github.com/4ian> I found a bug in my dialogue extension.
> Will you be ok if I include it in this pr?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/4ian/GDevelop/pull/1185?email_source=notifications&email_token=AAJYRAQCC3PKTDXMOKJXS53QEMKONA5CNFSM4IJZRQ32YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4G4K2Q#issuecomment-520996202>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAJYRASSZGGNWWVGPDN4JOLQEMKONANCNFSM4IJZRQ3Q>
> .
>

blurymind commented:
@4ian so far so good, the demo is almost ready. Before more people start using it, I want to try to make a more reusable code and see if this needs more work to be reusable.

Before it gets out, I wan't to see if we can get some of the names renamed, ad they don't seem to sound right
![2019-08-14_21-06](https://user-images.githubusercontent.com/6495061/63052616-b2e35600-bed7-11e9-8db6-dc047f4e293d.png)

1- "is running" - what is running? The only hint is the icon - but what if some people arent that good at reading the icon and what if at some point we get another extension with a similar icon. 
suggested change:  rename to "Dialogue is running" 
2, 3- I am not a big fan of the "repeat" for loop in the example for a couple of reasons:
- I have to create an index variable and manage it manually - is it possible to get the loop index without increment manually a variable
- The field to enter the number of repeat times has no autocompletion - so it is kind of hard to get that expression. I had to open the expression editor in another command, get it there and copy/paste

4. The use of has in some conditions seems good for a javascript variable naming convention, but for plain english it seems a bit odd to put the words in that order.
suggested change: rename to "Clipped text has completed scrolling" and "Selected option has changed"

Anyway, would appreciate getting your thoughts on this :) I want to get this right before way more people start using it.

blurymind commented:
I noticed a way to crash bondagejs :D, so now I think of tackling it in either yarn or bondage
https://github.com/hylyh/bondage.js/issues/50

4ian commented:
> suggested change: rename to "Dialogue is running"
> rename to "Clipped text has completed scrolling" and "Selected option has changed"

Yes agree for all. GDevelop sentences in the events sheet should include the "subject" they are dealing with.

> The use of "has" in some conditions at the beginning of the expression seems good for a javascript variable naming convention, but for plain english it seems a bit odd to put the words in that order.

Yes, same remark, GDevelop sentences must be complete.

> - I am not a big fan of the "repeat" for loop in the example for a couple of reasons:

Not sure about this one, I think I don't have a better solution. Unless avoiding the need for a loop entirely but not sure how.

blurymind commented:
@4ian thank you. That is actually a great idea :) I will give it a try tonight. The simpler it is to use the better

blurymind commented:
> `An error happened when trying to access the dialogue branch:`
> 
> Also, in which case does this happen? Can we prevent this?

it's whenever bondagejs tries to access a dialogue branch that does not exist in the tree data or contains illegal characters

4ian commented:
In this case, can we add a check on the extension side before going further? I would rather have explicit error checking rather than catching exception because this can hide other problems.

blurymind commented:
> In this case, can we add a check on the extension side before going further? I would rather have explicit error checking rather than catching exception because this can hide other problems.

unfortunately in this case it won't be trivial. I  can add some custom logic to catch the most common case, which is not existing - for the illegal characters one it is something that might be fixed on bondagejs

blurymind commented:
> Should you add a `return`?

I will have to revert that code as it's not working properly. Due to the way yarn structures its data, this is not trivial to do there. Let me see if there is a clean way

4ian commented:
Ok, in all cases would be great to validate/handle as much potential errors as possible and have the `catch` for the other cases, but that should be exceptional cases and if there is no other way to verify errors before.

blurymind commented:
these are kind of rare and you have to do some elaborate steps to cause them. For example to trigger the non existent branch name access, I had to manually delete the branch node and save the file.

Ideally I want to add a warning about that in yarn itself, so the user will never be allowed to cause the error without knowingly saving a file with an error in it.

The try/catch are only used where I want to be 100% sure bondagejs itself has no way of crashing the game - so accessing any methods it has that may not have good catching of errors

4ian commented:
Ok fair enough then! :)

blurymind commented:
having yarn embedded now with bondage is allowing me to catch errors much more rapidly, as the iteration time is much faster. You can really feel a huge difference when the editor is under one roof with bondagejs

blurymind commented:
btw I added an action that renders the options with the cursor in a single event:
![2019-08-15_21-06](https://user-images.githubusercontent.com/6495061/63123359-b42c8580-bfa0-11e9-8845-3aa980bed9cf.png)

You can pass as parameter what you want to use as selection cursor- although you can not pass utf characters/emojis there- as it will crash gd. In case of more ellaborate implementation of that- there are the old methods still in place

no need for complicated repeat loop and manual index variable

4ian commented:
> btw I added an action that renders the options with the cursor in a single event:

Looks great! :) 

> although you can not pass utf characters/emojis ➡️ there- as it will crash gd

Is this crashing GD or the game/bondage.js?

> Should I keep both in the example file?

Let's keep only the simple one using the new expression. People that want something more complex will be able to dig and figure out by themselves how to use a repeat event I think :)

blurymind commented:
The emoji characters - when loading them from the json file - they seem to work flawlessly.
![2019-08-15_21-17](https://user-images.githubusercontent.com/6495061/63123962-1fc32280-bfa2-11e9-83a8-b41e2ff1dc24.png)

However if you try to enter emojis in the event sheet and save your project,close it, reopen it and play the game, gdevelop.js starts to misbehave. 
I can attach an example project on a new issue if you are interested? :) I assumed that entering symbols and characters as string parameters was never ok with gdevelop.js

4ian commented:
Would be interested yeah, I'll take a look :)

Bouh commented:
Issue / improvement needed :

- When Yarn editor is open i can't use GDevelop. Yarn editor is over GD on block him. It's counter productive for multiples monitor.

About Yarn system : 

It's normal Yarn stopping at each lines ?

- When i talk to "enemy2" he say in json file : 
"Welcome to the dialogue tree demo project. 🤓\nPress Up/Z to interact with the tree critters, or alternatively jump on their heads"

I can only see the first line in GD, i need press Z for next lines. 

This mean i can't write in one time multiple line with \n. 
I need "call next line" with Yarn for each lines :( 
Maybe we need a new action for write all text from node. not just one line.

Im waiting documentation for create something from scratch and understand each events.

blurymind commented:
Yes that is correct. New lines are basically a way for yarn or in this case
bondagejs to interpret when to start printing the next message. So if you
want to have one huge message printed at once, you can do that with the
commands available too, it's just not in the example. :) The new line
pauses are intended by design.

The editor locking gdevelop is just how gdevelop uses external editors at
the moment- as modal windows- intended design again. I can change that for
yarn or even the other editors if @4ian oks it :)  i can understand how it
can be a bit inconvenient when it locks gdevelop while being used

On Thu, Aug 15, 2019, 10:32 PM Bouh <notifications@github.com> wrote:

> Issue / improvement needed :
>
>    - When Yarn editor is open i can't use GDevelop. Yarn editor is over
>    GD on block him. It's counter productive for multiples monitor.
>
> About Yarn system :
>
> It's normal Yarn stopping at each lines ?
>
>    - When i talk to "enemy2" he say in json file :
>    "Welcome to the dialogue tree demo project. 🤓\nPress Up/Z to interact
>    with the tree critters, or alternatively jump on their heads"
>
> I can only see the first line in GD, i need press Z for next lines.
>
> This mean i can't write in one time multiple line with \n.
> I need "call next line" with Yarn for each lines :(
> Maybe we need a new action for write all text from node. not just one line.
>
> Im waiting documentation for create something from scratch and understand
> each events.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/4ian/GDevelop/pull/1185?email_source=notifications&email_token=ABRRWVLPWETSG2QKZ3FYOCTQEXDO5A5CNFSM4IJZRQ32YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4NB3VA#issuecomment-521805268>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABRRWVPZ3WTZQMQJOZIPJL3QEXDO5ANCNFSM4IJZRQ3Q>
> .
>

blurymind commented:
@Bouh thank you for the feedback.

I will look into addressing these points:
- make sure its easy to print out a message with new lines in it at once easily
- make sure it is easy to render choices vertically
- make the yarn editor window non modal (needs @4ian to ok it :) )

Please note that the example project i shared so far is just for testing - it's not presentable yet. If you want to you can play with it, change the level layout or stuff in the event sheet. My focus is still on how the extension works and yarn

4ian commented:
> make the yarn editor window non modal (needs @4ian to ok it :) )
(include those things in the example)

So I would be ok if we think carefully about what this implies:
- If the user goes back to GDevelop, close some editor/etc..., is this still working?
- Can you open multiple editors? Is this supported/should be supported? How can we prevent multiple editors to be opened for the same resource (which is probably something we want to avoid)?

blurymind commented:
@4ian not sure why its not letting me use a boolean there
![2019-08-16_20-38](https://user-images.githubusercontent.com/6495061/63193640-e48a2780-c065-11e9-8c78-d989da504476.png)

4ian commented:
Boolean are not supported in expressions for now - sorry about that!
You can work around this by making another expression with a different name
OptionsTextWithNewLine? Can actually make the whole thing clearer.

On Fri, Aug 16, 2019 at 12:39 PM Todor Imreorov <notifications@github.com>
wrote:

> @4ian <https://github.com/4ian> not sure why its not letting me use a
> boolean there
> [image: 2019-08-16_20-38]
> <https://user-images.githubusercontent.com/6495061/63193640-e48a2780-c065-11e9-8c78-d989da504476.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/4ian/GDevelop/pull/1185?email_source=notifications&email_token=AAJYRAVOVFPVKHQLJURQ5ELQE366BA5CNFSM4IJZRQ32YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4PQSVQ#issuecomment-522127702>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAJYRAXZ3TJJWWWC22DFDSDQE366BANCNFSM4IJZRQ3Q>
> .
>

blurymind commented:
horizontalOptionsList and verticalOptionsList?

blurymind commented:
just to be clear, the method .setDefaultValue(...) only works for numbers?
Is there a way to do it for strings? Wish I could give it a default cursor when the user forgets to put one

blurymind commented:
@Bouh for getting the raw text of a branch (newlines included), there is an expression for that already:
```
        'BranchText',
        _('Get the raw text of the current branch'),
```
the downside to it is that you cant use <<commands>> or [[options]] in that branch, also new lines are not treated as logical separators - everything will print and is treated as raw text when you use that expression

blurymind commented:
Playing with this using it for signs, I now realize it would be useful to have a command to abruptly end a running dialogue- in the case of walking away from a sign. Will add it :)

blurymind commented:
@4ian I am ready to start writing the wiki. The demo project is close to ready now too. How do I add a wiki page for my extension?

blurymind commented:
updated project example
[dialogueExample.zip](https://github.com/4ian/GDevelop/files/3512148/dialogueExample.zip)

it does avatarts too

blurymind commented:
updated the example:
[dialogueExample.zip](https://github.com/4ian/GDevelop/files/3513076/dialogueExample.zip)

This time including some more cool features of the parser (like conditional dialogue). Managed to also catch a couple of minor, but annoying bugs

4ian commented:
For making a page, just visit the URL and edit the page to create it: http://wiki.compilgames.net/doku.php/gdevelop5/all-features/dialogue-tree :) (read a few pointers about [writing on the wiki here](http://wiki.compilgames.net/doku.php/gdevelop5/community/contribute-to-the-wiki))

The example will be really nice!

blurymind commented:
@4ian I registered an account at the wiki as blurymind, I don't yet have permission to edit :)

Things to do before this is ready:
- add wiki article
- add editor names to metadata

4ian commented on newIDE/app/src/ResourcesList/LocalYarnBridge.js:
```diff
@@ -0,0 +1,52 @@
+      createOrUpdateResource(project, new gd.JsonResource(), resourceName);
+
+      // $FlowFixMe - TODO: There is an error here to be solved.
+      onChangesSaved([{}], resourceName);
```
Let's not introduce any FlowFixMe.
Here, you're calling the onChangesSaved, but the first argument is an array and you've added an empty object in the array.
Avoid empty objects, it does not make sense here, just pass an empty array `[]`.

blurymind replied:
wasn't the metadata object used there? If so then, for now we can leave adding the name of the editor as metadata?

blurymind replied:
ok I removed the fixflowme there :) we will introduce metadata for editor names at a later stage

4ian commented on newIDE/app/src/ResourcesList/ResourceSelector.js:
```diff
@@ -206,6 +205,19 @@ export default class ResourceSelector extends React.Component<Props, State> {
+        extraOptions: {
+          initialResourceMetadata,
+        },
+        onChangesSaved: (newResourceData, newResourceName) => {
```
By the way, this "newResourceData" that is not being used is a huge hint that we've not designed properly this function arguments. The second argument might be not useful actually - and instead the editors calling onChangesSaved should always have been passing an array of resources.

Not sure if you see the idea? Basically, onChangesSaved is either called with the first argument, and no second argument, or is called with the second argument and an empty array for the first. This is a code smell - something is wrong.

Not the biggest deal but would be interesting to give a thought at simplifying this.

blurymind replied:
It would be nice to simplify it, but might cause other editors to break. Can I do it when implementing bassoon tracker support? That should be a smaller task, since the extension will have much less features :)

blurymind replied:
@4ian one quick fix is to swap the array and the resourcename in 
`onChangesSaved: (newResourceData, newResourceName)` and make the array an optional second parameter. That way we don't need to pass an empty array at least?

blurymind replied:
done now :)

4ian commented on newIDE/app/public/external/yarn/yarn-main.js:
```diff
@@ -0,0 +1,75 @@
+      .toString();
+
+    yarn.data.loadData(receivedOptions.externalEditorData, 'json', true);
+    electronWindow.setTitle('Yarn in GD: ' + receivedOptions.resourcePath);
```
GD => GDevelop - let's not use acronyms inside GDevelop itself.
Also, prefer a dash for separating title and filename:
`'Yarn in GDevelop - ' + receivedOptions.resourcePath `

4ian replied:
Also, I see that you've entered "GDevelop Dialogue Tree Editor" in index.html - would be cool to make these titles consistent by choosing either one or the other but let's not use two that are differents.

blurymind commented:
@4ian I removed the flowfixme's, made the titles consistent and added the demo project to the examples folder. The `onChangesSaved` will need to be sorted out after I get back from travelling - 2 weeks from now. I need to carefully do that refactoring test it extensively as it affects three editors so far

4ian commented:
Great, thanks for the demo project! Let's merge this once you have time to look at onChangesSaved and add the editor metadata :)

4ian commented:
@blurymind When you're back and have some time, let me know how it goes about onChangesSaved. Should not be that hard to clean and once it's done we can merge this :)

blurymind commented:
@4ian simplified the onChangesSaved now- so it only take one parameter - the resources. Also removed the redundant now flowfixmes

blurymind commented:
@4ian sorry about this, I have to use travis to fix the flow stuff, since my local flow checking is broken for some reason

blurymind commented:
ah damn, I need to refactor piskel's way of passing data, since it uses the two extra parameters too :/

blurymind commented:
ok I think I got the flow right now, please check if you can find any regressions in piskel's data handling and ability to store layers

blurymind commented:
now I am getting flow checking to work locally, but for some reason it is now testing all extensions and giving me 9 errors from various extensions

4ian commented:
What are these errors?

4ian commented:
I've just ran `yarn flow` on master and there should not be any issue.
In case of doubt, stop flow (`npm run flow -- stop` or `yarn flow stop`), remove node_modules and reinstall and relaunch flow (`npm run flow` or `yarn flow`).

blurymind commented:
this is turning out to be a bit of a pain in the neck to refactor. Will need to revisit it on a fresh mind tomorrow :)

4ian commented:
Ok :) 
Take a step back and think about:

What do I want to pass back when changes are saved? What are the fields I'm passing? What do they contain?

then writing the flow types is the same thing, but written as code :) 
Try to get something working on all 3 editors and I can help with the flow type if you tested and everything works ;)

blurymind commented:
onChangesSaved in spriteList.js doesn't seem to ever be called :/

blurymind commented:
ok my build environment was getting bugged, leading to my frustrations

blurymind commented:
ok after re-running flow server and re-running localhost, the odd stuff is gone. Will test some more for regressions tomorrow

4ian commented:
Great ✌️

blurymind commented:
@4ian after testing it today I found an old bug in piskel and changing name update - fixed now :) also updated the other editors to have a consistent window title format with Yarn. Other than that seems all good now. TheonChangesSaved method now also takes a single parameter as requested and flow test is passing

4ian commented:
Great! Let me take a look at this. Don't worry about Semaphore CI, I need to update how libGD.js is loaded.

You might have just fixed a bug reported on discord btw:
https://discordapp.com/channels/258623956158906368/377571085643677696

Having this refactored to be consistent is going to be useful for keeping this simple in the future :)

blurymind commented:
@4ian 
https://media.discordapp.net/attachments/377571085643677696/620982315413405697/issue.gif?width=1242&height=642
is actually a different bug :) looks interesting. Never thought of using it like this. I can try tofix it, but it might have some side effect of piskel creating unintended copy of a frame that was used twice

4ian commented:
Let's keep this for another PR, better stabilize/make sure this one is working in all the cases (apart from this bug)

blurymind commented:
@4ian not sure why now both online tests are not working. Running flow locally I get no errors

4ian commented:
Look into the logs ;) https://travis-ci.org/4ian/GDevelop/jobs/583326987
It's actually a formatting issue in master, just fixed it.

blurymind commented:
I fixed a minor issue with pasting via menu when in electron. I am attaching a new build below:
[yarn.zip](https://github.com/4ian/GDevelop/files/3602375/yarn.zip)

blurymind commented:
@4ian did some fixes to yarn today. Here is another new build (44476e1)
[yarn.zip](https://github.com/4ian/GDevelop/files/3613118/yarn.zip)

They are mostly on support for writing in other languages than english. The spellcheck detection will now work with other languages

blurymind commented:
By default it has all major language dictionaries, which increases its unzipped size to 10 mb, but the version I am attaching has them stripped and only the english one included.

If it can't load a dictionary locally,it downloads it automatically.

Would you  like me to include the french one and other ones in the zip that would be used by GD? That means french spellchecking will work offline in yarn. It will only increase the size with about 1mb for french.

4ian commented:
> If it can't load a dictionary locally,it downloads it automatically.

In this case I would rather start by including only English :) It it's really asked by lots of people we can include more/most used languages.

I'll upload the yarn zip file :) And take another look a the whole PR.

4ian commented on newIDE/app/public/external/yarn/yarn-main.js:
```diff
@@ -0,0 +1,78 @@
+
+    yarn.data.loadData(receivedOptions.externalEditorData, 'json', true);
+    electronWindow.setTitle(
+      'GDevelop Dialogue Tree Editor --' + receivedOptions.resourcePath
```
Just one caret (`GDevelop Dialogue Tree Editor -`)

4ian commented on newIDE/app/public/external/yarn/yarn-main.js:
```diff
@@ -0,0 +1,78 @@
+      receivedOptions.projectPath + '/NewFile.json';
+  }
+
+  receivedData = receivedOptions;
```
A bit confused by this line. Sounds weird to attributed received "options" into received "data". Would there be a better naming?

blurymind replied:
you're right. Renamed it to receivedData

4ian commented:
I've uploaded yarn.zip here as yarn-editor.zip: https://github.com/4ian/GDevelop/releases/download/v5.0.0-beta80/yarn-editor.zip

In your zip the folder was named dist, I had to rename it to `yarn-editor` if I'm not mistaken.

blurymind commented:
@4ian Thank you, I will have a look at updating the download  script and
address the review notes :)

On Sun, Sep 15, 2019, 12:27 PM Florian Rival <notifications@github.com>
wrote:

> I've uploaded yarn.zip here as yarn-editor.zip:
> https://github.com/4ian/GDevelop/releases/download/v5.0.0-beta80/yarn-editor.zip
>
> In your zip the folder was named dist, I had to rename it to yarn-editor
> if I'm not mistaken.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/4ian/GDevelop/pull/1185?email_source=notifications&email_token=ABRRWVMTP5XGHUGIT4YTJJLQJYL25A5CNFSM4IJZRQ32YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD6XOZ4A#issuecomment-531557616>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABRRWVMMKDHWIJ5O4UYSPCLQJYL25ANCNFSM4IJZRQ3Q>
> .
>

blurymind commented:
It is now downloading yarn from https://github.com/4ian/GDevelop/releases/download/v5.0.0-beta80/yarn-editor.zip
also renamed those things as requested :) 
I used a terminal command to get the hash
sha256sum yarn-editor.zip

it's still giving a warning for some reason, but does download and unzip it

blurymind commented:
@4ian I noticed that your zipped file contains a new folder called _MACOSX

Is that intended? My code doesn't use it at all and it requires adding a .gitignore line

4ian commented on newIDE/app/public/external/yarn/yarn-main.js:
```diff
@@ -0,0 +1,77 @@
+    receivedData.resourcePath = receivedData.projectPath + '/NewFile.json';
+  }
+
+  receivedData = receivedData;
```
This is surely broken because `receivedData = receivedData` is not changing the `receivedData` outside of the function scope. The variable outside the function is said to be "shadowed" by the parameter.

`receivedData` is a good name for the parameter. But for the variable outside the function, why is that called "received"? Is it really it purpose? What does this thing contains? Try to answer to these questions to find a proper name :) 

For finding a good name, inspect what the variable contains and what it is used for. I can see it's passed to the event "yarn-changes-saved".  Then when I look at where it's used, it seems to be passed and then, in `ipcRenderer.on('yarn-changes-saved', [...]` to be... never used ^^

In this case, if you don't need it, remove it and pass `null` - if that make sense in the code :)

4ian replied:
(I understand this is a detail, but I want to be sure that the code handling the external editor is as simple as possible.)

blurymind replied:
> > I noticed that your zipped file contains a new folder called _MACOSX
> > Is that intended? My code doesn't use it at all and it requires adding a .gitignore line
> 
> You can re-create a proper zip without it - it's automatically added by macOS Very disturbing - it's hidden by the macOS Finder.
> 
> > it's still giving a warning for some reason, but does download and unzip it
> 
> Weird, are you using the exact same file as the downloaded one? Any change in any bit in the file will totally change the checksum :) I'll update it myself if needed.

The _macosx folder is in the zip file you uploaded. You might have to reupload it :)

4ian replied:
Yeah if you give me a proper updated zip (wihtout the macos folder) I'll update the one hosted.

blurymind replied:
removed the receivedData variable and passing null. You are right - it is redundant in this case

4ian commented:
> I noticed that your zipped file contains a new folder called _MACOSX
> Is that intended? My code doesn't use it at all and it requires adding a .gitignore line

You can re-create a proper zip without it - it's automatically added by macOS 😠Very disturbing - it's hidden by the macOS Finder. 

> it's still giving a warning for some reason, but does download and unzip it

Weird, are you using the exact same file as the downloaded one? Any change in any bit in the file will totally change the checksum :) I'll update it myself if needed.

4ian commented on Extensions/DialogueTree/JsExtension.js:
```diff
@@ -474,18 +538,18 @@ module.exports = {
       .addCondition(
-        'HasSelectedOptionChanged',
-        _('Has selected option changed'),
+        'SelectedOptionHasChanged',
```
Can we keep "HasSelectedOptionChanged"? Start by the verb for a condition (Is/Has)

4ian replied:
The reason is that the internal name of conditions should be like functions. A function returning a "boolean" precisely. In most languages, it's a good idea to start a function name that returns a boolean by the verb:
"hasThingChanged", "isThingActivated", "shouldComponentUpdate", etc...

blurymind replied:
changed now

4ian commented on Extensions/DialogueTree/JsExtension.js:
```diff
@@ -551,19 +615,19 @@ module.exports = {
       .addCondition(
-        'HasClippedTextScrollingCompleted',
-        _('Has clipped text scrolling completed'),
+        'ClippedTextScrollingHasCompleted',
```
Same, can we revert to HasClippedTextScrollingCompleted?

blurymind replied:
updated now

4ian commented:
I've fixed typo/grammar mistakes in JsExtension.js

There are 4 conditions which are named internally:
* 'Compare dialogue state variable',
* 'Branch has been visited before',
* 'Current branch contains Tag',
* 'Current branch title is',

It's a bit late because this is already in the released extension (so this is going to be a breaking change for games using it), but nevertheless we should change these to names that are PascalCase like the other conditions/actions.
It's important because this could one day be used for other purposes (like automatically generating calls to functions, etc...), so never put a plain english sentence here.

4ian commented on Extensions/DialogueTree/JsExtension.js:
```diff
@@ -283,7 +347,7 @@ module.exports = {
         'ClippedLineText',
         _('Get dialogue line text clipped'),
         _(
-          'Get dialogue line text clipped by the typewriter effect. Use the ScrollClippedText action to control the typewriter effect.'
```
I've fixed this, don't use internal names for description shown to users ;)

4ian commented on Extensions/DialogueTree/JsExtension.js:
```diff
@@ -422,14 +486,14 @@ module.exports = {
     extension
       .addCondition(
         'IsDialogueLineType',
-        _('The dialogue line type is'),
```
Don't use partial sentences for titles of conditions. Put directly the name of what is being compared.

4ian replied:
(resolve once you've read this, I just want to highlight my changes ;))

4ian commented on Extensions/DialogueTree/JsExtension.js:
```diff
@@ -496,7 +560,7 @@ module.exports = {
     extension
       .addCondition(
         'Current branch title is',
```
See my comment, we should never use these for an internal title. Sadly it's a bit late so this will break the games using it, but anyway we should change it. Better now than later. 
Condition names are important because, even if the user are not seeing them, they are stored in the game files (which can be read and even edited manually, so we should give them names like in a programming language). Choosing a proper name is important, because once it's done it's here forever.

4ian replied:
(It's also possible that in the future, conditions could be called from expressions, so these condition names *will* be shown to the user)

blurymind replied:
should be "Compare current branch title" ?

4ian replied:
No that should be "CurrentBranchTitle" :) See the other conditions for example, it's not a user facing string, but the internal name of the condition. Same for the others.

blurymind replied:
ah right, got confused there xD changing it now. Hopefully not many people have started using the extension yet

4ian commented on newIDE/app/public/external/jfxr/jfxr-main.js:
```diff
@@ -71,18 +71,29 @@ ipcRenderer.on('jfxr-open', (event, receivedOptions) => {
   });
+
+  electronWindow.setTitle(
+    'GDevelop Sound Effects Editor - ' +
```
Not sure if we should "hide" the name of the editor we're embedding. What do you think of at least keeping their original names like this: "Sound Effects Editor (Jfxr) - "?

blurymind replied:
You're right - I need to re-add the names. That way the any editor specific bugs will be more likely to be submitted to the editor's github, not gdevelop's. We shouldn't mislead the users in any way that these editors are a part of GD

blurymind commented:
Ah that's strange. I zipped in Linux. Sorry for the confusion. Not sure why
it added that folder. Will try to apply the new review notes tomorrow

On Sun, Sep 15, 2019, 10:49 PM Florian Rival <notifications@github.com>
wrote:

> *@4ian* commented on this pull request.
> ------------------------------
>
> In newIDE/app/public/external/yarn/yarn-main.js
> <https://github.com/4ian/GDevelop/pull/1185#discussion_r324482759>:
>
> > +    receivedData.externalEditorData = fs
> +      .readFileSync(receivedData.resourcePath, 'utf8')
> +      .toString();
> +
> +    yarn.data.loadData(receivedData.externalEditorData, 'json', true);
> +    electronWindow.setTitle(
> +      'GDevelop Dialogue Tree Editor - ' + receivedData.resourcePath
> +    );
> +
> +    pathEditorHeader.toggle();
> +  } else {
> +    // If GD has sent no path, we need to set one for yarn
> +    receivedData.resourcePath = receivedData.projectPath + '/NewFile.json';
> +  }
> +
> +  receivedData = receivedData;
>
> Yeah if you give me a proper updated zip (wihtout the macos folder) I'll
> update the one hosted.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/4ian/GDevelop/pull/1185?email_source=notifications&email_token=ABRRWVJCQDZO6J42SCANEULQJ2UXDA5CNFSM4IJZRQ32YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCEYGD6A#discussion_r324482759>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABRRWVJT3FKXE2OF5F43MJTQJ2UXDANCNFSM4IJZRQ3Q>
> .
>

4ian commented:
I used your zip, but had to unzip it, rename the folder "dist" to "yarn-editor" and zip it again (but with macos, hence the extra folder). If you do it by yourself, ensuring to have the folder named yarn-editor, and send here the zip, I can host it directly without changing anything :)

Cool thanks, this is pretty huge PR so I want to double check everything and I discover things when I look again. Pretty important to get stuff properly done because it's here for a long time :) And polishing work is important to get a quality extension.

blurymind commented:
I see, apologies for the inconvenience. I forgot to rename the folder before zipping it. Will upload a new one later today :)  also remove that redundant object I was passing. You are right- it is not required in this case, since I am not writing any metadata from yarn to GD

blurymind commented:
@4ian here is an updated yarn package:
[yarn-editor.zip](https://github.com/4ian/GDevelop/files/3616326/yarn-editor.zip)
 (44476e1 )

4ian commented:
I updated the yarn-editor.zip package hosted on GitHub :)

blurymind commented:
@4ian thank you for catching these in the last moment. I have now updated the PR and applied all notes.
I have also updated the yarn-dialogue example project to work with the renamed commands

blurymind commented:
ah have I missed some, let me check again

blurymind commented:
updated jsextension.js now and it should hopefully have no other badly named parameters now

blurymind commented:
not sure why my sha256 key is still wrong :/

4ian commented:
Right this is the hash of the folder indeed - I fixed it. You can "cheat" by `console.log`-ing the value of the computed hash in `scripts/import-zipped-editor.js` :)

This should be ready for some final reviews I guess?

4ian commented on Extensions/DialogueTree/dialoguetools.js:
```diff
@@ -218,9 +284,16 @@ gdjs.dialogueTree.confirmSelectOption = function() {
+      gdjs.dialogueTree.goToNextDialogueLine();
+    } catch (error) {
+      console.error(
+        `An error happened when trying to access the dialogue branch!`,
```
What could cause this error?

blurymind replied:
Selecting an option that doesn't exists in the yarn data tree will cause bondagejs to crash the game. This is why I have added a try catch, so instead nothing happens and a warning is logged :) Ideally this should never happen. Yarn automatically creates nodes to options you make a new link to. But for flexibility it is still possible for the user to delete a node and forget to erase a link to it. I think of later on adding a warning before saving the file for such dead node links

blurymind replied:
Thanks to this catch, the player will still be capable of trying to select another option in the game that does have a node linked to it - and the game will proceed to  work.

4ian replied:
Sounds good then!

4ian commented on newIDE/app/public/external/piskel/piskel-main.js:
```diff
@@ -365,31 +364,18 @@ ipcRenderer.on('piskel-load-animation', (event, receivedOptions) => {
 
   // Load a custom save file(s) header
   const pathEditorHeaderDiv = document.getElementById('path-editor-header');
-  const headerStyle = {
```
Removing these hardcoded styles is definitely a nice thing 👍

blurymind replied:
They were an eye sore 😔 sorry I ever put them there. Was a pain in the neck to check the styling and reuse it

4ian commented on newIDE/app/public/external/yarn/yarn-main.js:
```diff
@@ -0,0 +1,75 @@
+
+let yarn = null;
+
+const saveAndClose = pathEditor => {
```
Let's rename this variable pathEditor to pathEditorHeader, to be consistent with the naming below (`const pathEditorHeader = createPathEditorHeader`)

4ian commented on newIDE/app/src/ResourcesList/LocalYarnBridge.js:
```diff
@@ -0,0 +1,49 @@
+  ipcRenderer.removeAllListeners('yarn-changes-saved');
+  ipcRenderer.on(
+    'yarn-changes-saved',
+    (event, newFilePath, externalEditorData) => {
```
If `externalEditorData` is not used, remove it from the arguments list.

4ian commented on newIDE/electron-app/app/main.js:
```diff
@@ -190,6 +190,31 @@ app.on('ready', function() {
+    });
+  });
+
+  ipcMain.on('yarn-changes-saved', (event, newFilePath, externalEditorData) => {
```
Seems like externalEditorData is not used, remove it from the `.on` and the `send`

4ian commented on newIDE/app/public/external/yarn/yarn-main.js:
```diff
@@ -0,0 +1,75 @@
+const saveAndClose = pathEditor => {
+  const savePath = pathEditor.state.fullPath;
+  yarn.data.saveTo(savePath, yarn.data.getSaveData('json'), () => {
+    ipcRenderer.send('yarn-changes-saved', savePath, null);
```
The third argument (null) is never used, remove it entirely, otherwise it implies that it's used somewhere else for some purpose.

4ian commented on newIDE/app/src/ResourcesList/LocalYarnBridge.js:
```diff
@@ -0,0 +1,49 @@
+    resourceNames[0]
+  );
+
+  const externalEditorData = {
```
Let's not change it now, but I'm not a fan of calling an object externalEditorData and have inside something called... externalEditorData. Definitely a code smell.
Let's not change it for now but could have been named differently, like "yarnArguments" or something like this.

blurymind replied:
ok, will leave it for another pr :)

4ian commented:
Still a few comments and should be good to go!

blurymind commented:
> Still a few comments and should be good to go!
@4ian 
Thank you for looking into these. I will get them sorted tomorrow :)

blurymind commented:
[yarn-editor.zip](https://github.com/4ian/GDevelop/files/3627667/yarn-editor.zip) (a5fb76b )
@4ian sorry to ask you about this again, but can you update yarn-editor again? I recently fixed a bug where it is underlying a word as misspelled if the word is followed by a punctuation sign.
https://github.com/YarnSpinnerTool/YarnEditor/commit/a5fb76b771481c92b2ec4af47d4da460c4b56aa8

Also btw the git repository has moved to
https://github.com/YarnSpinnerTool/YarnEditor
due to the owner of the initial one passing away recently

4ian commented:
Update the zip and the hash. I've seen that the zip is now 3mb while it used to be a bit less than 1mb. By looking quickly, I see that the vendor.xxx.js file which used to be ~1mb is now 3mb. Any idea why? A different compression utility? Or a new dependency added recently?

blurymind commented:
Ah sorry about that. Its probably a change in the build. I will get it down
again

On Wed, Sep 18, 2019, 9:44 PM Florian Rival <notifications@github.com>
wrote:

> Update the zip and the hash. I've seen that the zip is now 3mb while it
> used to be a bit less than 1mb. By looking quickly, I see that the
> vendor.xxx.js file which used to be ~1mb is now 3mb. Any idea why? A
> different compression utility? Or a new dependency added recently?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/4ian/GDevelop/pull/1185?email_source=notifications&email_token=ABRRWVP45QYFPMLEE7PCSQDQKKHJBA5CNFSM4IJZRQ32YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7BMMHY#issuecomment-532858399>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABRRWVORF3FWFUCHEZUAHIDQKKHJBANCNFSM4IJZRQ3Q>
> .
>

blurymind commented:
@4ian I think by mistake I packaged a non optimised dev build. Here is fixed one:

 
[yarn-editor.zip](https://github.com/4ian/GDevelop/files/3628421/yarn-editor.zip)
(190f5d1)

blurymind commented:
@4ian I updated yarn's styling to better work on smartphones recently, so I had to update the label of the save and close button to just say "apply" to be shorter. Hope you don't mind. Also made yarn's window bigger, so more content can be seen. Might as well when it's modal

4ian commented:
Both seems fine to me :) 

Online version works well! Good job :)

blurymind commented:
> Both seems fine to me :)
> 
> Online version works well! Good job :)

Thank you 😸 
I plan to add dropbox i/o to it after I sort out the rest of the android specific little issues. That way people will be able to work on dialogues on their smartphones, just by visiting the webapp.
One thing that got me abit stuck is figuring out how to best hide the address bar when its on a smartphone, since vertical space is so precious

4ian commented:
You can offer a button to go fullscreen: 
https://developers.google.com/web/fundamentals/native-hardware/fullscreen/
https://www.sitepoint.com/use-html5-full-screen-api/ 
https://developer.mozilla.org/en-US/docs/Web/API/Fullscreen_API

I was thinking of doing this too in GDevelop when the viewport width or height is small enough.

blurymind commented:
> You can offer a button to go fullscreen:
> https://developers.google.com/web/fundamentals/native-hardware/fullscreen/
> https://www.sitepoint.com/use-html5-full-screen-api/
> https://developer.mozilla.org/en-US/docs/Web/API/Fullscreen_API
> 
> I was thinking of doing this too in GDevelop when the viewport width or height is small enough.

Do you think it could be possible to automatically trigger it without a button? I have a way of detecting when on android now. Anyways, sorry to go off a tangent :)

blurymind commented:
I managed to get it to work as a progressive web app, but it needs a lot more work to be mobile friendly :)  The touch input, caching locally to work offline, support for dropbox i/o.. In any case the full screen thing is now solved :D Thank you for the links

blurymind commented:
@4ian here is an updated build with the new dropbox features:
[yarn-editor.zip](https://github.com/4ian/GDevelop/files/3639884/yarn-editor.zip) ( e5a2786 )
Sorry I keep updating it. Will stop working on it for a while :)
In any case let me know if this PR needs anything else

blurymind commented:
It would be best if for now the dropbox feature is disabled for electron - as it doesnt work there. I was wondering if its worth implementing for desktop anyways. Here is a newer builddisabling dropbox if the app is running on electron
[yarn-editor.zip](https://github.com/4ian/GDevelop/files/3640060/yarn-editor.zip)
(a82626)

4ian commented:
It's fine to have this disabled, I'm thinking of adding something like this to GD later but this probably need a bit of rethinking about the way resources are loaded/handled because GD will need to get access to a lot of files at the same time from Dropbox/Google Drive or another provider.

I'll update the zip

4ian commented on newIDE/electron-app/app/main.js:
```diff
@@ -12,12 +12,12 @@ const autoUpdater = require('electron-updater').autoUpdater;
 const {
   uploadGameFolderToBucket,
-  uploadArchiveToBucket,
+  uploadArchiveToBucket
```
This file has all the comma removed, probably because you runned Prettier with a different trailingComma setup: https://prettier.io/docs/en/configuration.html

Can you fix these changes? they should not be included in the PR.

blurymind replied:
ah, sorry. I need to update my prettier setting

4ian commented:
Updated :)
On my side I don't have anything else related to this PR apart from some trailing comma that have been removed in a file - always double check the files in the PR after committing to avoid including changes that are unrelated to the PR and "pollute" it ;)

blurymind commented:
@4ian I fixed the removed trailing commas now :) :+1:

4ian commented:
Sounds good, let me take a final look and merge this tomorrow :)

On Mon, 23 Sep 2019, 18:26 Todor Imreorov, <notifications@github.com> wrote:

> @4ian <https://github.com/4ian> I fixed the removed trailing commas now
> :) 👍
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/4ian/GDevelop/pull/1185?email_source=notifications&email_token=AAJYRASK7XP6TDUTPZGPC23QLD32VA5CNFSM4IJZRQ32YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7LT73I#issuecomment-534200301>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAJYRATDIL7G4FVRYR6NDQ3QLD32VANCNFSM4IJZRQ3Q>
> .
>

4ian commented:
Sorry merged, too early :/

I have this issue when trying the example, it does not seem to work:
<img width="804" alt="Screenshot 2019-09-24 23 28 24" src="https://user-images.githubusercontent.com/1280130/65555055-4a01eb80-df23-11e9-981a-7d51936dac8f.png">

Can you check again and clean the example? Thanks :) Continue working on your branch and open a new PR with it

blurymind commented:
That's strange. Does it get stuck there? On the event sheet are there any
unknown events?

On Tue, Sep 24, 2019, 11:31 PM Florian Rival <notifications@github.com>
wrote:

> Sorry merged, too early :/
>
> I have this issue when trying the example, it does not seem to work:
> [image: Screenshot 2019-09-24 23 28 24]
> <https://user-images.githubusercontent.com/1280130/65555055-4a01eb80-df23-11e9-981a-7d51936dac8f.png>
>
> Can you check again and clean the example? Thanks :) Continue working on
> your branch and open a new PR with it
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/4ian/GDevelop/pull/1185?email_source=notifications&email_token=ABRRWVIV25NTZNHGUFHWLVTQLKILDA5CNFSM4IJZRQ32YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7QABJA#issuecomment-534773924>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABRRWVMVI2N2EIXW2AKGW3DQLKILDANCNFSM4IJZRQ3Q>
> .
>

blurymind commented:
I renamed extension commands as requested, updated them in the demo too..
but I didn't run the script for updating the demos. Just updated in the
examples folder. Could that be causing it for you to not work?

On Wed, Sep 25, 2019, 8:15 AM Todor Imreorov <blurymind@gmail.com> wrote:

> That's strange. Does it get stuck there? On the event sheet are there any
> unknown events?
>
> On Tue, Sep 24, 2019, 11:31 PM Florian Rival <notifications@github.com>
> wrote:
>
>> Sorry merged, too early :/
>>
>> I have this issue when trying the example, it does not seem to work:
>> [image: Screenshot 2019-09-24 23 28 24]
>> <https://user-images.githubusercontent.com/1280130/65555055-4a01eb80-df23-11e9-981a-7d51936dac8f.png>
>>
>> Can you check again and clean the example? Thanks :) Continue working on
>> your branch and open a new PR with it
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/4ian/GDevelop/pull/1185?email_source=notifications&email_token=ABRRWVIV25NTZNHGUFHWLVTQLKILDA5CNFSM4IJZRQ32YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7QABJA#issuecomment-534773924>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ABRRWVMVI2N2EIXW2AKGW3DQLKILDANCNFSM4IJZRQ3Q>
>> .
>>
>

4ian commented:
It look as if some extensions have been installed and now missing: 
<img width="955" alt="Screenshot 2019-09-25 08 42 34" src="https://user-images.githubusercontent.com/1280130/65579866-80198c80-df70-11e9-8fa1-e24d81d3250a.png">
<img width="959" alt="Screenshot 2019-09-25 08 42 44" src="https://user-images.githubusercontent.com/1280130/65579867-80198c80-df70-11e9-8e89-a5275e2212e6.png">

It seems I can't do anything in the demo, I can't move or do anything with the keyboard or mouse, might be related?

blurymind commented:
Ah I know what's causing this. I experimented with the visual extensions
feature, creating some event sheet extensions. But then I deleted them.
Strangely the demo continued to work for me, but breaks as an example. I
guess the fix would be to remove any reference leftovers to them in the
project

On Wed, Sep 25, 2019, 8:43 AM Florian Rival <notifications@github.com>
wrote:

> It look as if some extensions have been installed and now missing:
> [image: Screenshot 2019-09-25 08 42 34]
> <https://user-images.githubusercontent.com/1280130/65579866-80198c80-df70-11e9-8fa1-e24d81d3250a.png>
> [image: Screenshot 2019-09-25 08 42 44]
> <https://user-images.githubusercontent.com/1280130/65579867-80198c80-df70-11e9-8e89-a5275e2212e6.png>
>
> It seems I can't do anything in the demo, I can't move or do anything with
> the keyboard or mouse, might be related?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/4ian/GDevelop/pull/1185?email_source=notifications&email_token=ABRRWVKCZ5JJRB2M5ZLS3D3QLMJCZA5CNFSM4IJZRQ32YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7Q6ALQ#issuecomment-534896686>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABRRWVNWG5UUONVTPNDZ3WLQLMJCZANCNFSM4IJZRQ3Q>
> .
>

4ian commented:
You can remove them from the behaviors tab in GD. Can you take a look at this to double check that the example is working once all objects have their reference to these behaviors removed? I suspect there might be something else, because in theory a deleted behavior like this should not make a game unplayable (unless you have events relying on them).

blurymind commented:
> You can remove them from the behaviors tab in GD. Can you take a look at this to double check that the example is working once all objects have their reference to these behaviors removed? I suspect there might be something else, because in theory a deleted behavior like this should not make a game unplayable (unless you have events relying on them).

I need to carefully inspect the project. The strange thing is that it was running fine for me on linux/dev environment. I tested it yesterday

4ian commented:
Ok double check and if it works I'll have another look. I wonder if it's related to some changes merged in master since you started. Notably, the JSON are now by default preloaded at the beginning of the game (this can be disabled in the Resources tab, if you have a lot of JSON for dialogs and are fine with loading them during the game).

blurymind commented:
> JSON are now by default preloaded at the beginning of the game (this can be disabled in the Resources tab, if you have a lot of JSON for dialogs and are fine with loading them during the game).

That is also a likely reason for it. I haven't merged dev into it for a while and the extension demo does rely on loading the resource from a json resource. This is sort of at the root of the feature

blurymind commented:
@4ian this is very weird. I merged master into the pr branch and the example still works fine for me. Are you pressing the z button to advance the dialogue?

4ian commented:
Huh no :p I tried to press every button but I guess I missed the z button. Would be interesting to either add a "Press z to continue" to something like this because it's not obvious and pretty sure people will make the same mistake :)
Can you see about cleaning the example from the bad behaviors though?

blurymind commented:
> Sorry merged, too early :/
> 
> I have this issue when trying the example, it does not seem to work:
> <img alt="Screenshot 2019-09-24 23 28 24" width="804" src="https://user-images.githubusercontent.com/1280130/65555055-4a01eb80-df23-11e9-981a-7d51936dac8f.png">
> 
> Can you check again and clean the example? Thanks :) Continue working on your branch and open a new PR with it

here it is :)
https://github.com/4ian/GDevelop/pull/1236
